libraries/botbuilder-dialogs-adaptive-testing/src/mocks/mockLuisRecognizer.ts

```diff
@@ -0,0 +1,166 @@
+import { Activity, RecognizerResult } from 'botbuilder-core';
+import {
+    HttpRequestMessage,
+    MockHttpRequestMiddleware,
+    MockHttpRequestMiddlewareKey,
+} from './mockHttpRequestMiddleware';
+import { HttpResponseMessage } from '../httpRequestMocks/httpResponseMock';
+import { DynamicList } from 'botbuilder-ai/lib/dynamicList';
+import { ListElement } from 'botbuilder-ai/lib/listElement';
+import { ExternalEntity } from 'botbuilder-ai/lib/externalEntity';
+
+/**
+ * Wrap murmurhash v3.
+ *
+ * @param {string} source Source string to compute hash.
+ * @returns {number} Computed hash result.
+ */
+const stableHash = (source: string): number => murmurhash.murmur3(source);
+
+/**
+ * Compute result of x XOR y.
+ *
+ * @param {number} x First parameter.
+ * @param {number} y Second parameter.
+ * @returns {number} Computed XOR result.
+ */
+function xor(x: number, y: number): number {
+    const bufferX = new ArrayBuffer(4);
+    new DataView(bufferX).setUint32(0, x);
+    const bufferY = new ArrayBuffer(4);
+    new DataView(bufferY).setUint32(0, y);
+    const result = buffer.xor(Buffer.from(bufferX), Buffer.from(bufferY));
+    return parseInt(result.toString('hex'), 16);
+}
+
+/**
+ * Test class for creating cached LUIS responses for testing.
+ */
+export class MockLuisRecognizer extends Recognizer {
+    private _responseDir: string;
+    private _recognizer: LuisAdaptiveRecognizer;
+
+    /**
+     * Initializes a new instance of the MockLuisRecognizer class.
+     *
+     * @param {LuisAdaptiveRecognizer} recognizer LUIS recognizer definition.
+     * @param {string} resourceDir Where the settings file generated by lubuild is found.
+     * @param {string} name Name of the LUIS model
+     */
+    public constructor(recognizer: LuisAdaptiveRecognizer, resourceDir: string, name: string) {
+        super();
+        this._recognizer = recognizer;
+        this._responseDir = path.join(resourceDir, 'cachedResponses', name);
+        if (!fs.existsSync(this._responseDir)) {
+            fs.mkdirSync(this._responseDir);
+        }
+    }
+
+    public async recognize(
+        dialogContext: DialogContext,
+        activity: Activity,
+        telemetryProperties?: Record<string, string>,
+        telemetryMetrics?: Record<string, number>
+    ): Promise<RecognizerResult> {
+        const options = this._recognizer.recognizerOptions(dialogContext);
+        options.includeAPIResults = true;
+        const middleware: MockHttpRequestMiddleware = dialogContext.context.turnState.get(MockHttpRequestMiddlewareKey);
+        middleware?.setFallback((request: HttpRequestMessage) => this._fallback(request, activity.text, options));
+        const result = await this._recognizer.recognize(dialogContext, activity, telemetryProperties, telemetryMetrics);
+        middleware?.setFallback(undefined);
+        return result;
+    }
+
+    private _responsePath(utterance: string, options: LuisRecognizerOptionsV3): string {
```

Comment:
A small description of the general gist of this code would be helpful. What's the desired output string? Is it a hash that represents a unique API request? Is there any reason we can't use `crypto` to compute a checksum?

Reply:
Yes, the desired output is a unique file path representing utterance and options of a LUIS query, of course it is ok to use `crypto` to compute a checksum but I have to port this method from C#: https://github.com/microsoft/botbuilder-dotnet/blob/072cb74dea90e814239fa5b7da0d5494263a2203/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/Mocks/MockLuisRecognizer.cs#L83 to compute exactly the same hash value to leverage existing response file in the C# repo.

Reply:
Ah - that makes sense! Fair enough.